core/blockchain.go

```diff
@@ -3158,6 +3219,147 @@ func (bc *BlockChain) IsEnablePruneBeaconChainFeature() bool {
+	}
+
+	currentBlockNum := currentBlock.NumberU64()
+	if currentBlockNum < newBlkNum {
+		start := time.Now()
+		for i := currentBlockNum; i <= newBlkNum; i++ {
+			blk := bc.GetBlockByNumber(i)
+			if blk == nil {
+				// todo check later
+				return nil
+			}
+
+			err := bc.tikvFastForward(blk, logs)
+			if err != nil {
+				return err
+			}
+		}
+		log.Printf("sync to %d use %v", currentBlockNum, time.Now().Sub(start))
+	}
+
+	return nil
+}
+
+// tikvCleanCache used for tikv mode, clean block tire data from redis
+func (bc *BlockChain) tikvCleanCache() {
+	var count int
+	for to := range bc.cleanCacheChan {
+		for i := bc.latestCleanCacheNum + 1; i <= to; i++ {
+			// build previous block statedb
+			fromBlock := bc.GetBlockByNumber(i)
+			fromTrie, err := state.New(fromBlock.Root(), bc.stateCache)
+			if err != nil {
+				continue
+			}
+
+			// build current block statedb
+			toBlock := bc.GetBlockByNumber(i + 1)
+			toTrie, err := state.New(toBlock.Root(), bc.stateCache)
+			if err != nil {
+				continue
+			}
+
+			// diff two statedb and delete redis cache
+			start := time.Now()
+			count, err = fromTrie.DiffAndCleanCache(bc.ShardID(), toTrie)
+			if err != nil {
+				utils.Logger().Warn().
+					Err(err).
+					Msg("tikv clean cache error")
+				break
+			}
+
+			log.Println("tikvCleanCache block", i, "count", count, "use", time.Now().Sub(start))
+			bc.latestCleanCacheNum = i
+		}
+	}
+}
+
+// tikvPreemptMaster used for tikv mode, writer node need preempt master or come be a follower
+func (bc *BlockChain) tikvPreemptMaster(fromBlock, toBlock uint64) bool {
+	for {
+		// preempt master
+		if ok, _ := bc.redisPreempt.TryLock(60); ok {
+			return true
+		}
+
+		// follower
+		if bc.CurrentBlock().NumberU64() >= toBlock {
+			return false
+		}
+
+		time.Sleep(time.Second)
+	}
+}
+
+// rangeBlock get the block range of blocks
+func (bc *BlockChain) rangeBlock(blocks types.Blocks) (uint64, uint64) {
+	if len(blocks) == 0 {
+		return 0, 0
+	}
+	max := blocks[0].NumberU64()
+	min := max
+	for _, tmpBlock := range blocks {
+		if tmpBlock.NumberU64() > max {
+			max = tmpBlock.NumberU64()
+		} else if tmpBlock.NumberU64() < min {
+			min = tmpBlock.NumberU64()
+		}
+	}
+
+	return min, max
+}
+
+// RedisPreempt used for tikv mode, get the redis preempt instance
+func (bc *BlockChain) RedisPreempt() *redis_helper.RedisPreempt {
+	return bc.redisPreempt
+}
+
+// InitTiKV used for tikv mode, init the tikv mode
+func (bc *BlockChain) InitTiKV(conf *harmonyconfig.TiKVConfig) {
+	bc.cleanCacheChan = make(chan uint64, 10)
+
+	if conf.Role == "Writer" {
+		// only writer need preempt permission
+		bc.redisPreempt = redis_helper.CreatePreempt(fmt.Sprintf("shard_%d_preempt", bc.ShardID()))
+	}
+
+	if conf.Debug {
+		// used for init redis cache
+		// If redis is empty, the hit rate will be too low and the synchronization block speed will be slow
+		// set LOAD_PRE_FETCH is yes can significantly improve this.
+		if os.Getenv("LOAD_PRE_FETCH") == "yes" {
```

Comment:
why an environment variable here?  why not a configuration option?

Reply:
This is only used when starting the tikv cluster for the first time